File: packages/shared-types/src/shock.ts
/**
 * Shock types for modifying demographic rates.
 * Shocks represent events that alter mortality, fertility, or migration.
 */

import type { Gender } from './demographic';

/**
 * Type of demographic rate affected by the shock.
 */
export type ShockType = 'mortality' | 'fertility' | 'migration';

/**
 * Age range for targeting shocks.
 */
export interface AgeRange {
  /** Minimum age (inclusive) */
  min: number;
  /** Maximum age (inclusive) */
  max: number;
}

/**
 * Modifier types for rate adjustments.
 */
export interface MultiplierModifier {
  type: 'multiplier';
  value: number;
}

export interface AbsoluteModifier {
  type: 'absolute';
  value: number;
}

export interface FunctionModifier {
  type: 'function';
  expression: string;
}

export type Modifier = MultiplierModifier | AbsoluteModifier | FunctionModifier;

/**
 * Shock modifier definition.
 * Applies a multiplier to demographic rates during a specified period.
 */
export interface Shock {
  /** Unique shock identifier */
  id: string;
  /** Human-readable name */
  name: string;
  /** Optional description */
  description?: string;
  /** Type of rate to modify */
  type: ShockType;
  /** First year of shock (inclusive) */
  startYear: number;
  /** Last year of shock (inclusive) */
  endYear: number;
  /** Target regions ('all' or specific region IDs) */
  targetRegions: 'all' | string[];
  /** Target genders ('all' or specific genders) */
  targetGenders: 'all' | Gender[];
  /** Target age range ('all' or specific range) */
  targetAges: 'all' | AgeRange;
  /** Rate modifier */
  modifier: Modifier | number;
  /**
   * Optional intensity curve over time.
   * If provided, modifier is scaled by this curve.
   * Array index corresponds to year offset from startYear.
   * Values should be 0-1 (0 = no effect, 1 = full modifier).
   */
  intensityCurve?: number[];
}

/**
 * Predefined shock templates for common scenarios.
 */
export interface ShockTemplate {
  /** Template identifier */
  id: string;
  /** Template name */
  name: string;
  /** Template description */
  description: string;
  /** Category for UI grouping */
  category: 'pandemic' | 'war' | 'migration' | 'policy' | 'custom';
  /** Default shock configuration (without years) */
  defaults: Omit<Shock, 'id' | 'startYear' | 'endYear'>;
}

/**
 * Common shock templates.
 */
export const SHOCK_TEMPLATES: ShockTemplate[] = [
  {
    id: 'pandemic-moderate',
    name: 'Moderate Pandemic',
    description: 'Increased mortality for elderly population (similar to severe flu season)',
    category: 'pandemic',
    defaults: {
      name: 'Moderate Pandemic',
      type: 'mortality',
      targetRegions: 'all',
      targetGenders: 'all',
      targetAges: { min: 65, max: 120 },
      modifier: 1.3, // 30% increase
    },
  },
  {
    id: 'pandemic-severe',
    name: 'Severe Pandemic',
    description: 'Significantly increased mortality across all ages, especially elderly',
    category: 'pandemic',
    defaults: {
      name: 'Severe Pandemic',
      type: 'mortality',
      targetRegions: 'all',
      targetGenders: 'all',
      targetAges: { min: 0, max: 120 },
      modifier: 1.5, // 50% increase overall
      intensityCurve: [0.3, 1.0, 0.8, 0.4, 0.1], // Peaks in year 2
    },
  },
  {
    id: 'war-conventional',
    name: 'Conventional War',
    description: 'Increased mortality for young adult males + emigration',
    category: 'war',
    defaults: {
      name: 'War Casualties',
      type: 'mortality',
      targetRegions: 'all',
      targetGenders: ['male'],
      targetAges: { min: 18, max: 45 },
      modifier: 2.0, // 100% increase
    },
  },
  {
    id: 'baby-boom',
    name: 'Baby Boom',
    description: 'Increased fertility rates (post-war, economic prosperity)',
    category: 'policy',
    defaults: {
      name: 'Baby Boom',
      type: 'fertility',
      targetRegions: 'all',
      targetGenders: ['female'],
      targetAges: { min: 20, max: 40 },
      modifier: 1.4, // 40% increase
    },
  },
  {
    id: 'fertility-decline',
    name: 'Fertility Decline',
    description: 'Decreased fertility (economic crisis, urbanization)',
    category: 'policy',
    defaults: {
      name: 'Fertility Decline',
      type: 'fertility',
      targetRegions: 'all',
      targetGenders: ['female'],
      targetAges: { min: 15, max: 49 },
      modifier: 0.7, // 30% decrease
    },
  },
  {
    id: 'mass-emigration',
    name: 'Mass Emigration',
    description: 'Large-scale emigration of working-age population',
    category: 'migration',
    defaults: {
      name: 'Mass Emigration',
      type: 'migration',
      targetRegions: 'all',
      targetGenders: 'all',
      targetAges: { min: 20, max: 40 },
      modifier: -50000, // 50,000 emigrants per year
    },
  },
  {
    id: 'refugee-influx',
    name: 'Refugee Influx',
    description: 'Large-scale immigration of refugees (all ages)',
    category: 'migration',
    defaults: {
      name: 'Refugee Influx',
      type: 'migration',
      targetRegions: 'all',
      targetGenders: 'all',
      targetAges: 'all',
      modifier: 100000, // 100,000 immigrants per year
    },
  },
];

// ============================================================
// MODIFIER HELPERS
// ============================================================

/**
 * Create a multiplier modifier.
 */
export function multiplier(value: number): MultiplierModifier {
  return { type: 'multiplier', value };
}

/**
 * Create an absolute modifier.
 */
export function absolute(value: number): AbsoluteModifier {
  return { type: 'absolute', value };
}

/**
 * Create an additive modifier for migration shocks (alias for absolute).
 */
export function additive(value: number): AbsoluteModifier {
  return absolute(value);
}

/**
 * Apply a modifier to a base value.
 */
export function applyModifier(baseValue: number, modifier: Modifier): number {
  switch (modifier.type) {
    case 'multiplier':
      return baseValue * modifier.value;
    case 'absolute':
      return baseValue + modifier.value;
    case 'function':
      throw new Error('Function modifiers must be evaluated on the backend');
  }
}

// ============================================================
// SHOCK APPLICABILITY
// ============================================================

/**
 * Check if a shock applies to a specific cohort at a given time.
 */
export function shockApplies(
  shock: Shock,
  year: number,
  age: number,
  gender: Gender,
  regionId: string
): boolean {
  // Check year range
  if (year < shock.startYear || year > shock.endYear) {
    return false;
  }

  // Check region
  if (shock.targetRegions !== 'all' && !shock.targetRegions.includes(regionId)) {
    return false;
  }

  // Check gender
  if (shock.targetGenders !== 'all' && !shock.targetGenders.includes(gender)) {
    return false;
  }

  // Check age
  if (shock.targetAges !== 'all') {
    if (age < shock.targetAges.min || age > shock.targetAges.max) {
      return false;
    }
  }

  return true;
}

// ============================================================
// SHOCK FACTORY FUNCTIONS
// ============================================================

interface PandemicShockParams {
  id: string;
  name: string;
  startYear: number;
  endYear: number;
  mortalityIncrease: number;
  minAge?: number;
  description?: string;
}

/**
 * Create a pandemic shock.
 */
export function pandemicShock(params: PandemicShockParams): Shock {
  return {
    id: params.id,
    name: params.name,
    description: params.description ?? `Pandemic affecting population aged ${params.minAge ?? 0}+`,
    type: 'mortality',
    startYear: params.startYear,
    endYear: params.endYear,
    targetRegions: 'all',
    targetGenders: 'all',
    targetAges: { min: params.minAge ?? 0, max: 120 },
    modifier: multiplier(params.mortalityIncrease),
  };
}

interface WarShockParams {
  id: string;
  name: string;
  startYear: number;
  endYear: number;
  mortalityIncrease: number;
  description?: string;
}

/**
 * Create a war shock (targeting young males).
 */
export function warShock(params: WarShockParams): Shock {
  return {
    id: params.id,
    name: params.name,
    description: params.description ?? 'War casualties primarily affecting young adult males',
    type: 'mortality',
    startYear: params.startYear,
    endYear: params.endYear,
    targetRegions: 'all',
    targetGenders: ['male'],
    targetAges: { min: 18, max: 45 },
    modifier: multiplier(params.mortalityIncrease),
  };
}

interface MigrationCrisisShockParams {
  id: string;
  name: string;
  startYear: number;
  endYear: number;
  netMigration: number;
  description?: string;
}

/**
 * Create a migration crisis shock.
 */
export function migrationCrisisShock(params: MigrationCrisisShockParams): Shock {
  const direction = params.netMigration >= 0 ? 'immigration' : 'emigration';
  const count = Math.abs(params.netMigration).toLocaleString();
  
  return {
    id: params.id,
    name: params.name,
    description: params.description ?? `Net ${direction} of ${count} people per year`,
    type: 'migration',
    startYear: params.startYear,
    endYear: params.endYear,
    targetRegions: 'all',
    targetGenders: 'all',
    targetAges: 'all',
    modifier: absolute(params.netMigration),
  };
}

// ============================================================
// TEMPLATE HELPERS
// ============================================================

/**
 * Helper to create a shock from a template.
 */
export function createShockFromTemplate(
  template: ShockTemplate,
  startYear: number,
  endYear: number,
  overrides?: Partial<Shock>
): Shock {
  return {
    id: crypto.randomUUID(),
    startYear,
    endYear,
    ...template.defaults,
    ...overrides,
  };
}

File: worker/src/types/shock.rs
//! Shock types for demographic rate modifiers

use serde::{Deserialize, Serialize};
use super::{AgeGroup, Gender};

/// Type of demographic shock
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ShockType {
    Mortality,
    Fertility,
    Migration,
}

/// Target specification for shocks
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Target<T> {
    All,
    Specific(Vec<T>),
}

impl<T> Target<T> {
    pub fn is_all(&self) -> bool {
        matches!(self, Target::All)
    }
}

/// Age target for shocks
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AgeTarget {
    #[serde(rename = "all")]
    All,
    Range(AgeGroup),
}

impl AgeTarget {
    pub fn contains(&self, age: u32) -> bool {
        match self {
            AgeTarget::All => true,
            AgeTarget::Range(group) => group.contains(age),
        }
    }
}

/// Shock modifier types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ShockModifier {
    /// Multiply the base value (1.5 = 50% increase)
    Multiplier { value: f64 },
    /// Add an absolute value
    Absolute { value: f64 },
    /// Custom function (evaluated at runtime)
    Function { expression: String },
}

impl ShockModifier {
    pub fn multiplier(value: f64) -> Self {
        Self::Multiplier { value }
    }

    pub fn absolute(value: f64) -> Self {
        Self::Absolute { value }
    }

    /// Apply the modifier to a base value
    pub fn apply(&self, base_value: f64) -> f64 {
        match self {
            ShockModifier::Multiplier { value } => base_value * value,
            ShockModifier::Absolute { value } => base_value + value,
            ShockModifier::Function { .. } => {
                // Function modifiers would require an expression evaluator
                // For now, just return the base value
                base_value
            }
        }
    }
}

/// Shock: Modifier applied to demographic rates
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Shock {
    pub id: String,
    pub name: String,
    pub description: String,
    #[serde(rename = "type")]
    pub shock_type: ShockType,
    pub start_year: u32,
    pub end_year: u32,
    pub target_regions: Target<String>,
    pub target_genders: Target<Gender>,
    pub target_ages: AgeTarget,
    pub modifier: ShockModifier,
}

impl Shock {
    /// Check if the shock applies to the given parameters
    pub fn applies(
        &self,
        year: u32,
        age: u32,
        gender: Gender,
        region_id: &str,
    ) -> bool {
        // Check year range
        if year < self.start_year || year > self.end_year {
            return false;
        }

        // Check region
        if let Target::Specific(regions) = &self.target_regions {
            if !regions.iter().any(|r| r == region_id) {
                return false;
            }
        }

        // Check gender
        if let Target::Specific(genders) = &self.target_genders {
            if !genders.contains(&gender) {
                return false;
            }
        }

        // Check age
        if !self.target_ages.contains(age) {
            return false;
        }

        true
    }
}

/// Create a pandemic shock template
pub fn pandemic_shock(
    id: &str,
    name: &str,
    start_year: u32,
    end_year: u32,
    mortality_increase: f64,
    min_age: Option<u32>,
) -> Shock {
    Shock {
        id: id.to_string(),
        name: name.to_string(),
        description: format!(
            "Pandemic: {}% mortality increase for ages {}+",
            ((mortality_increase - 1.0) * 100.0).round(),
            min_age.unwrap_or(0)
        ),
        shock_type: ShockType::Mortality,
        start_year,
        end_year,
        target_regions: Target::All,
        target_genders: Target::All,
        target_ages: min_age
            .map(|min| AgeTarget::Range(AgeGroup::range(min, 120)))
            .unwrap_or(AgeTarget::All),
        modifier: ShockModifier::multiplier(mortality_increase),
    }
}

/// Create a war shock template
pub fn war_shock(
    id: &str,
    name: &str,
    start_year: u32,
    end_year: u32,
    mortality_increase: f64,
    min_age: Option<u32>,
    max_age: Option<u32>,
) -> Shock {
    let min = min_age.unwrap_or(18);
    let max = max_age.unwrap_or(45);
    
    Shock {
        id: id.to_string(),
        name: name.to_string(),
        description: format!(
            "War: {}% mortality increase for males {}-{}",
            ((mortality_increase - 1.0) * 100.0).round(),
            min,
            max
        ),
        shock_type: ShockType::Mortality,
        start_year,
        end_year,
        target_regions: Target::All,
        target_genders: Target::Specific(vec![Gender::Male]),
        target_ages: AgeTarget::Range(AgeGroup::range(min, max)),
        modifier: ShockModifier::multiplier(mortality_increase),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_shock_modifier_multiplier() {
        let modifier = ShockModifier::multiplier(1.5);
        assert_eq!(modifier.apply(100.0), 150.0);
    }

    #[test]
    fn test_shock_modifier_absolute() {
        let modifier = ShockModifier::absolute(50.0);
        assert_eq!(modifier.apply(100.0), 150.0);
    }

    #[test]
    fn test_shock_applies() {
        let shock = pandemic_shock("test", "Test", 2025, 2026, 1.5, Some(65));
        
        // Should apply
        assert!(shock.applies(2025, 70, Gender::Male, "CZ"));
        assert!(shock.applies(2025, 70, Gender::Female, "CZ"));
        
        // Should not apply
        assert!(!shock.applies(2024, 70, Gender::Male, "CZ")); // Wrong year
        assert!(!shock.applies(2025, 50, Gender::Male, "CZ")); // Wrong age
    }

    #[test]
    fn test_war_shock_targets_males_only() {
        let shock = war_shock("war", "War", 2025, 2026, 3.0, None, None);
        
        assert!(shock.applies(2025, 30, Gender::Male, "CZ"));
        assert!(!shock.applies(2025, 30, Gender::Female, "CZ"));
    }
}

